theano/tensor/nnet/tests/test_corr.py

```diff
@@ -0,0 +1,313 @@
+        s = 1.
+        orig_image_data = image_data
+        if border_mode is not 'full':
+            s = -1.
+        out_shape2d = (numpy.array(N_image_shape[-2:]) +
+                       s * numpy.array(N_filter_shape[-2:]) - s)
+        out_shape2d = numpy.ceil(out_shape2d / numpy.array(subsample))
+        out_shape = (N_image_shape[0], N_filter_shape[0]) + tuple(out_shape2d)
+        ref_output = numpy.zeros(out_shape)
+
+        # loop over output feature maps
+        ref_output.fill(0)
+        if border_mode == 'full':
+            image_data2 = numpy.zeros((N_image_shape[0], N_image_shape[1],
+                                       N_image_shape[2] + 2 * N_filter_shape[2] - 2,
+                                       N_image_shape[3] + 2 * N_filter_shape[3] - 2))
+            image_data2[:, :, N_filter_shape[2] - 1:N_filter_shape[2] - 1 + N_image_shape[2],
+                        N_filter_shape[3] - 1:N_filter_shape[3] - 1 + N_image_shape[3]] = image_data
+            image_data = image_data2
+            N_image_shape = image_data.shape
+        for bb in range(N_image_shape[0]):
+            for nn in range(N_filter_shape[0]):
+                for im0 in range(N_image_shape[1]):
+                    filter2d = filter_data_corr[nn, im0, :, :]
+                    image2d = image_data[bb, im0, :, :]
+                    for row in range(ref_output.shape[2]):
+                        irow = row * subsample[0]  # image row
+                        for col in range(ref_output.shape[3]):
+                            icol = col * subsample[1]  # image col
+                            ref_output[bb, nn, row, col] += (image2d[
+                                irow:irow + N_filter_shape[2],
+                                icol:icol + N_filter_shape[3]] * filter2d[::-1, ::-1]
+                            ).sum()
+
+        self.assertTrue(_allclose(theano_output, ref_output))
+
+        # TEST GRADIENT
+        if verify_grad:
+            utt.verify_grad(sym_CorrMM, [orig_image_data, filter_data])
+
+    def test_basic(self):
+        """
+        Tests that basic correlations work for odd and even
+        dimensions of image and filter shapes, as well as rectangular
+        images and filters.
+        """
+        self.validate((2, 2, 3, 3), (2, 2, 2, 2), 'valid', verify_grad=False)
+        self.validate((3, 2, 8, 8), (4, 2, 5, 5), 'valid', verify_grad=False)
+        self.validate((3, 2, 7, 5), (5, 2, 2, 3), 'valid')
+        self.validate((3, 2, 7, 5), (5, 2, 3, 2), 'valid', verify_grad=False)
+        self.validate((3, 2, 8, 8), (4, 2, 5, 5), 'full', verify_grad=False)
+        self.validate((3, 2, 7, 5), (5, 2, 2, 3), 'full')
+        self.validate((1, 10, 213, 129), (46, 10, 212, 1), 'valid',
+                      verify_grad=False)
+
+    def test_img_kernel_same_shape(self):
+        self.validate((3, 2, 3, 3), (4, 2, 3, 3), 'full')
+        self.validate((3, 2, 3, 3), (4, 2, 3, 3), 'valid')
+
+    @attr('slow')
```

Comment:
Can you find that test? If it take less then 1s you can remove the slow,
otherwise keep it.
Le 9 sept. 2015 14:03, "Jesse Livezey" <notifications@github.com> a écrit :

> In theano/tensor/nnet/tests/test_corr.py
> <https://github.com/Theano/Theano/pull/3285#discussion_r39075818>:
>
> > +        images and filters.
> > +        """
> > +        self.validate((2, 2, 3, 3), (2, 2, 2, 2), 'valid', verify_grad=False)
> > +        self.validate((3, 2, 8, 8), (4, 2, 5, 5), 'valid', verify_grad=False)
> > +        self.validate((3, 2, 7, 5), (5, 2, 2, 3), 'valid')
> > +        self.validate((3, 2, 7, 5), (5, 2, 3, 2), 'valid', verify_grad=False)
> > +        self.validate((3, 2, 8, 8), (4, 2, 5, 5), 'full', verify_grad=False)
> > +        self.validate((3, 2, 7, 5), (5, 2, 2, 3), 'full')
> > +        self.validate((1, 10, 213, 129), (46, 10, 212, 1), 'valid',
> > +                      verify_grad=False)
> > +
> > +    def test_img_kernel_same_shape(self):
> > +        self.validate((3, 2, 3, 3), (4, 2, 3, 3), 'full')
> > +        self.validate((3, 2, 3, 3), (4, 2, 3, 3), 'valid')
> > +
> > +    @attr('slow')
>
> @nouiz <https://github.com/nouiz>, should I remove the 'slow'? I'm not
> sure what this does. Subsampling with this implementation should be faster.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Theano/Theano/pull/3285/files#r39075818>.
>